core/lib/spree/transaction.rb

```diff
@@ -0,0 +1,154 @@
+# frozen_string_literal: true
+
+require "spree/registry"
+
+module Spree
+  module Transaction
+    def self.included(klass)
+      klass.include(self[])
+    end
+
+    def self.[](registry: Spree::OperationRegistry, db: true)
+      Class.new(Module) do
+        def initialize(registry:, db:)
+          @registry = registry
+          @db = db
+        end
+
+        def included(klass)
+          klass.extend(ClassMethods)
+          class_exec(@registry, @db) do |registry, db|
+            klass.define_method(:initialize) do |**kwargs|
+              @registry = registry.merge(**kwargs)
+              @db = db
+            end
+          end
+
+          klass.define_method(:call) do |*args, **kwargs|
+            result = nil
+
+            (@db ? ActiveRecord::Base.method(:transaction) : method(:yield_self)).call do
+              catch(:halt) do
+                transaction = klass.instance_variable_get(:@block)
+                unwrapped_result = Execution.new(registry: @registry).instance_exec(*args, **kwargs, &transaction)
+                Spree::Result.success(unwrapped_result)
+              end.tap do |wrapped_result|
+                result = wrapped_result
+                raise ActiveRecord::Rollback if result.failure? && @db
+              end
+            end
+
+            result
+          end
+        end
+      end.new(registry: registry, db: db)
+    end
+
+    module ClassMethods
+      def transaction(db: true, &block)
+        @block = block
+        @db = db
```

Comment:
I wonder if the `db: true` parameter isn't a smell this should actually be two methods in the future - a db `transaction` and a block of aggregated operation within a service look different to me. 

The first is atomic, ie can be treated as a single operation, while the second is not atomic in all scenarios - it would be highly dependent on the operations chained; anything with side-effects that is not the last call would break this.

Am I overthinking this? 😅

Reply:
Hey, @cesartalves, thanks for looking into it 🙌 

I think both `db: true` & `db: false` should be considered the same in terms of atomicity. In a purely functional programming language, that would be the case. However, in Ruby (and most other languages, FWIW), we're free to do whatever we want. For instance, it's also true that you can open an `ActiveRecord::Base.transaction` block, create a file within, and roll it back. The file will still exist after rolling back, so there we had a side effect. The `transaction` block opens a business transaction, and the `db` parameter can make it also a database transaction.

I agree that a boolean parameter is a smell most of the time, as the caller already knows which path to take. Ideally, we could have two separate methods, like `transaction` & `transaction_with_db`. However, given how implicit database transactions are in Rails, I think it's better to default to the safe side. We could also have `transaction` & `transaction_without_db`, but it looks ugly 🙂 I'm open to discussing it further.